tests/fv/flannel_migration_test.go

```diff
@@ -0,0 +1,381 @@
+// Copyright (c) 2019 Tigera, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package fv_test
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"os"
+	"os/exec"
+	"regexp"
+	"strings"
+	"time"
+
+	"github.com/sirupsen/logrus"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+
+	gocidr "github.com/apparentlymart/go-cidr/cidr"
+	"github.com/projectcalico/felix/fv/containers"
+	"github.com/projectcalico/kube-controllers/tests/testutils"
+	"github.com/projectcalico/libcalico-go/lib/apiconfig"
+	api "github.com/projectcalico/libcalico-go/lib/apis/v3"
+	backend "github.com/projectcalico/libcalico-go/lib/backend/api"
+	"github.com/projectcalico/libcalico-go/lib/backend/model"
+	client "github.com/projectcalico/libcalico-go/lib/clientv3"
+	"github.com/projectcalico/libcalico-go/lib/ipam"
+	cnet "github.com/projectcalico/libcalico-go/lib/net"
+	"github.com/projectcalico/libcalico-go/lib/options"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+const (
+	// Flannel migration controller will run on this node.
+	controllerNodeName = "node-0"
+	// Master node Name.
+	masterNodeName = "node-1"
+
+	flannelDs = "kube-flannel-ds-amd64"
+
+	flannelSubnetEnv = "FLANNEL_NETWORK=192.168.0.0/16;FLANNEL_SUBNET=192.168.1.1/24;FLANNEL_MTU=8951;FLANNEL_IPMASQ=true;"
+)
+
+var (
+	emptyLabel = map[string]string{}
+)
+
+var _ = Describe("flannel-migration-controller FV test", func() {
+	var (
+		etcd                *containers.Container
+		migrationController *containers.Container
+		apiserver           *containers.Container
+		calicoClient        client.Interface
+		bc                  backend.Client
+		k8sClient           *kubernetes.Clientset
+		controllerManager   *containers.Container
+		kconfigfile         *os.File
+		err                 error
+		flannelCluster      *testutils.FlannelCluster
+	)
+
+	logKubectl := func(args ...string) {
+		out, err := apiserver.ExecOutput(args...)
+		Expect(err).ShouldNot(HaveOccurred())
+		logrus.Infof("--- kubectl output --- \n%s", out)
+	}
+
+	startController := func() {
+		// Add 3 seconds delay before main thread starts, this is to make sure FV can add watch channels
+		// before controller logs out to stderr.
+		// (container.Run returns after 'docker ps' shows the container, the polling interval is 1 second.)
+		// Add 60 seconds delay before main thread exits, this is to make sure controller is still running
```

Comment:
I don't really understand this comment - we don't seem to need this for our other FV tests for kube-controllers. Maybe we should chat about this more so I understand.

Reply:
kube-controllers focus more on the consistency between k8s resources and calico resources. migration-controller is more about applying correct steps to existing k8s resources.  Most of the migration controller test cases rely on watching logs to validate the workflow of the controller. It is not perfect but probably is the most efficient way to test different flows, error handlings etc. without adding more complicated code to output those flows to some external storage and do validations. 

So a typical test case for migration-controller would look like this
```
1. c := container.Run(`migration-controller with env`)
2. w := c.AddWatchStderr(Regex("target string"))
3. Eventually(w is closed) at timeout of 20 seconds.
4. c.Stop()
```
The first problem we have is `container.Run` will return after the container is listed by `docker ps`. The felix FV code polling status of `docker ps` at a 1 second interval. So it is possible that at the time `container.Run` returns, the migration-controller has been running for 1 seconds and exits. Adding a watch after container is stopped does not work. So I added 3 seconds delay so we can add watch before the migration-controller logs out anything. 

Adding a 60 seconds delay before exits is to make enough container is still running and be stopped correctly after `Eventually` timeout. 

We could improve Felix FV code by providing interface to do something like 
```
1. c := container.Create(`migration-controller with env`)
2. w := c. AddWatchStderr(Regex("target string"))
3. c.Start()
4. Eventually(w is closed) at timeout of 20 seconds.
4. c.Stop()
```
However, I thought it would be too risky to make the change since we are so close for a major release.